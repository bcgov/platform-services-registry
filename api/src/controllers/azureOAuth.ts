import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
