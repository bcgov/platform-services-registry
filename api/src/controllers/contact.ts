//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
