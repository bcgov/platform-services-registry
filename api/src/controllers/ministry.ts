//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from "@bcgov/common-nodejs-utils";
import { Request, Response } from "express";
import DataManager from "../db";
import shared from "../libs/shared";

const dm = new DataManager(shared.pgPool);

const fetchMinistrySponsors = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  try {
    const results = await MinistryModel.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = "Unable fetch ministries";
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export default fetchMinistrySponsors;
