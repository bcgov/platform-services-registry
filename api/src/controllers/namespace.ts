//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;
  const aBody = { ...body, profileId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(201).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { namespaceId } = params;

  try {
    const results = await NamespaceModel.findById(Number(namespaceId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;
  const { name, clusterId } = body;
  const aBody = { name, profileId, clusterId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    await NamespaceModel.update(namespaceId, aBody);

    res.status(204).end();
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
