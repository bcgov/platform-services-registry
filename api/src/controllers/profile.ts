//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import shared from '../libs/shared';
import { isNotAuthorized, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserId(user.id);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    res.status(200).json(record);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix. If it exists, this
  // overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();

    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    categoryId,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
  } = body;

  try {
    const record = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      categoryId,
      busOrgId,
      prioritySystem,
      userId: record.userId,
      namespacePrefix: record.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
    };

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(profileId);

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

    res.status(202).end();
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ContactModel, ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const results = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
