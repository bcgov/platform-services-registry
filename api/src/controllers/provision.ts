//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { CLUSTER_NAMES, GOLD_QUORUM_COUNT } from '../constants';
import DataManager from '../db';
import { ProjectProfile } from '../db/model/profile';
import { RequestEditType } from '../db/model/request';
import { AuthenticatedUser } from '../libs/authmware';
import { fetchBotMessageRequests } from '../libs/bot-message';
import { MessageType, sendProvisioningMessage } from '../libs/messaging';
import { createNamespaces } from '../libs/namespace';
import { getProvisionStatus, updateProvisionStatus } from '../libs/profile';
import { processProfileContactsEdit, processProfileQuotaSizeEdit, processProjectProfileEdit } from '../libs/request';
import shared from '../libs/shared';
import { generateNamespaceNames } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const provisionProfileNamespaces = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { profileId } = params;
  const clusters = body;
  const { ProfileModel, NamespaceModel, ClusterModel } = dm;

  try {
    const existing = await NamespaceModel.findForProfile(profileId);
    if (existing.length === 0) {
      const profile = await ProfileModel.findById(profileId);

      if (!profile) {
        const errmsg = 'Unable to fetch info for provisioning';
        throw new Error(errmsg);
      }

      const namespaceNames = await generateNamespaceNames(profile.namespacePrefix);

      const namespaces = await createNamespaces(namespaceNames, profileId);

      for (const cluster of clusters) {
        const clusterDetails = await ClusterModel.findByName(cluster);
        await NamespaceModel.createProjectSet(Number(clusterDetails.id), namespaces);
      }
    }
    res.status(202).end();
  } catch (err) {
    const message = `Unable to provision namespaces for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const provisionerCallbackHandler = async (
  { body }: { body: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { prefix, clusterName } = body;

  try {
    const profile = await ProfileModel.findByPrefix(prefix);
    if (!profile) {
      throw new Error(`Cant find any profile for the given prefix ${prefix}`);
    }

    if (!CLUSTER_NAMES.includes(clusterName)) {
      throw new Error(`Unknown cluster name: ${clusterName} included in callback response`);
    }

    const isProfileProvisioned = await getProvisionStatus(profile);

    if (isProfileProvisioned) {
      await processProvisionedProfileEditRequest(profile, clusterName);
    } else {
      await updateProvisionedProfile(profile, clusterName);
    }
    res.status(204).end();
  } catch (err) {
    const message = `Unable to handle provisioner callback for profile prefix ${prefix}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

const updateProvisionedProfile = async (profile: ProjectProfile, clusterName: string): Promise<void> => {
  const { RequestModel } = dm;
  try {
    // Must check if all bot_messages are complete before completing.
    // Step 1. fetch active bot_messages for request

    const requests = await RequestModel.findForProfile(Number(profile.id));
    const request = requests.pop();
    if (!request) {
      return;
    }

    const botMessageSet = await fetchBotMessageRequests(Number(request.id))

    if (botMessageSet.length !== GOLD_QUORUM_COUNT) {
      await updateProvisionStatus(profile, true);

      logger.info(`Sending CHES message (${MessageType.ProvisioningCompleted}) for ${profile.id}`);
      await sendProvisioningMessage(Number(profile.id), MessageType.ProvisioningCompleted);
      logger.info(`CHES message sent for ${profile.id}`);

      await RequestModel.updateCompletionStatus(Number(request.id));
    }

    const botMessage = botMessageSet.filter(message => message.clusterName === clusterName).pop()
    if (!botMessage) {
      const errmsg = `Unable to get bot message with cluster name: ${clusterName}`;
      throw new Error(errmsg);
    }
    await RequestModel.updateCallbackStatus(Number(botMessage.id))

    return;
  } catch (err) {
    const message = `Unable to update provisioned profile`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const processProvisionedProfileEditRequest = async (profile: ProjectProfile, clusterName: string): Promise<void> => {
  const { RequestModel } = dm;
  try {
    if (!profile.id) {
      throw new Error('Cant read the given profileId');
    }

    const requests = await RequestModel.findForProfile(profile.id);
    const request = requests.pop();
    if (!request) {
      return;
    }

    const botMessageSet = await fetchBotMessageRequests(Number(request.id))

    if (botMessageSet.length !== GOLD_QUORUM_COUNT) {
      switch (request.editType) {
        case RequestEditType.ProjectProfile:
          await processProjectProfileEdit(request);
          break;
        case RequestEditType.Contacts:
          await processProfileContactsEdit(request);
          break;
        case RequestEditType.QuotaSize:
          await processProfileQuotaSizeEdit(request);
          break;
        default:
          throw new Error(`Invalid edit type for request ${request.id}`);
      }
      await RequestModel.updateCompletionStatus(Number(request.id));
    }

    const botMessage = botMessageSet.filter(message => message.clusterName === clusterName).pop()
    if (!botMessage) {
      const errmsg = `Unable to get bot message with cluster name: ${clusterName}`;
      throw new Error(errmsg);
    }
    await RequestModel.updateCallbackStatus(Number(botMessage.id))
  } catch (err) {
    const message = `Unable to process provisioned profile edit request`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};
