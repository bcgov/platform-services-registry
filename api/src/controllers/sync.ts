//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Request, Response } from 'express';
import DataManager from '../db';
import { ProjectProfile } from '../db/model/profile';
import { RequestEditType, RequestType } from '../db/model/request';
import { contextForEditing, contextForProvisioning } from '../libs/fulfillment';
import { getProvisionStatus } from '../libs/profile';
import shared from '../libs/shared';
import { replaceForDescription } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const getAllProvisionedProfileIds = async (
  req: Request, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const clusterName: any = req.query['cluster-name'];

  if (!(clusterName === undefined || typeof clusterName === 'string')) {
    throw errorWithCode('Unable to determine the provided cluster name', 400);
  }

  try {
    let profiles: ProjectProfile[] = await ProfileModel.findAll();
    if (clusterName !== undefined) {
      profiles = await filterProfilesBySelectedClusterName(profiles, clusterName);
    }

    const provisionedProfileIds: number[] = [];
    for (const profile of profiles) {
      const isProvisioned: boolean = await getProvisionStatus(profile);
      if (isProvisioned && profile.id) {
        provisionedProfileIds.push(profile.id);
      }
    }

    res.status(200).json(provisionedProfileIds);
  } catch (err) {
    const message = 'Unable to fetch all provisioned profile ids';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const getProvisionedProfileBotJson = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { profileId } = params;
  const { ProfileModel } = dm;

  try {
    const profile = await ProfileModel.findById(Number(profileId));
    if (!profile) {
      const errmsg = `Cant find any profile for the given profile ${profileId}`;
      throw new Error(errmsg);
    }

    const result = await getProvisionStatus(profile);
    if (!result) {
      const errmsg = `This profile ${profileId} is not provisioned`;
      throw new Error(errmsg);
    }

    const context = await contextForProvisioning(profileId, true);

    res.status(200).json(replaceForDescription(context));
  } catch (err) {
    const message = `Unable to get provisioned profile bot json for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const getAllProfileIdsUnderPending = async (
  req: Request, res: Response
): Promise<void> => {
  const clusterName: any = req.query['cluster-name'];

  if (!(clusterName === undefined || typeof clusterName === 'string')) {
    throw errorWithCode('Unable to determine the provided cluster name', 400);
  }

  try {
    let profiles: ProjectProfile[] = await getProfilesUnderPendingEditOrCreate();
    if (clusterName !== undefined) {
      profiles = await filterProfilesBySelectedClusterName(profiles, clusterName);
    }

    res.status(200).json(profiles.map(p => p.id));
  } catch (err) {
    const message = 'Unable fetch all profile ids that are under pending edit';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const getProfileBotJsonUnderPending = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { profileId } = params;
  const { RequestModel } = dm;

  try {
    const profiles: ProjectProfile[] = await getProfilesUnderPendingEditOrCreate();
    const profileIds: (number | undefined)[] = profiles.map(p => p.id);
    if (!profileIds.includes(Number(profileId))) {
      const errmsg = `This profile is not under any pending edit / create request`;
      throw new Error(errmsg);
    }

    let context;

    const requests = await RequestModel.findForProfile(profileId);
    // if the queried profile is under pending edit
    if (requests.length > 0) {
      const request = requests.pop();
      if (!request) {
        const errmsg = `Unable to get request`;
        throw new Error(errmsg);
      }
      switch (request.type) {
        case RequestType.Create:
          context = await contextForProvisioning(profileId, false);
          break;
        case RequestType.Edit:
          if (!request.editType){
            const errmsg = `Unable to get request edit Type`;
            throw new Error(errmsg);
          }
          if (request.editType === RequestEditType.ProjectProfile) {
            request.editObject = JSON.stringify(request.editObject)
          }

          context = await contextForEditing(profileId, request.editType, request.editObject);
          break;
        default:
          throw new Error(`Invalid type for request ${request.id}`);
        }
    }
    res.status(200).json(replaceForDescription(context));
  } catch (err) {
    const message = `Unable to get profile (currently under pending edit / create) bot json
    for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

const getProfilesUnderPendingEditOrCreate = async (): Promise<ProjectProfile[]> => {
  const { RequestModel, ProfileModel } = dm;
  try {
    // process those profiles that are under pending EDIT
    const requests = await RequestModel.findActiveByFilter('requires_human_action', false );
    const profileUnderRequestPromises: Promise<ProjectProfile>[] =
      requests.map((request) => ProfileModel.findById(request.profileId));

    const profilesUnderPendingEdit: ProjectProfile[] = await Promise.all(profileUnderRequestPromises);

    // process those profiles that are under pending CREATE
    const profilesUnderPendingCreate: ProjectProfile[] = [];

    const profiles: ProjectProfile[] = await ProfileModel.findAll();
    for (const profile of profiles) {
      const isProvisioned = await getProvisionStatus(profile);
      if (!isProvisioned) {
        profilesUnderPendingCreate.push(profile);
      }
    }

    return profilesUnderPendingEdit.concat(profilesUnderPendingCreate);
  } catch (err) {
    const message = 'Unable to get a list of profiles for those that are under pending edit / create';
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const filterProfilesBySelectedClusterName = async (profiles: ProjectProfile[], clusterName: string):
  Promise<ProjectProfile[]> => {
  try {
    return profiles.filter((profile: ProjectProfile) => profile.primaryClusterName === clusterName);

  } catch (err) {
    const message = 'Unable to filter a list of profiles by selected cluster name';
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};
