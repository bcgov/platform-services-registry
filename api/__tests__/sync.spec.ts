//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { getAllProvisionedProfileIds, getProvisionedProfileBotJson } from '../src/controllers/sync';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profile.json');
const selectProfile = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/select-profile-namespaces.json');
const selectProfileNamespaces = JSON.parse(fs.readFileSync(p1, 'utf8'));

const client = new Pool().connect();

jest.mock('../src/libs/namespace-set', () => {
  const p2 = path.join(__dirname, 'fixtures/select-default-cluster.json');
  const selectDefaultCluster = JSON.parse(fs.readFileSync(p2, 'utf8'));

  return {
    isNamespaceSetProvisioned: jest.fn().mockReturnValue(true),
    getDefaultCluster: jest.fn().mockReturnValue(selectDefaultCluster),
  };
});

jest.mock('../src/libs/fulfillment', () => {
  const p3 = path.join(__dirname, 'fixtures/provisioning-context.json');
  const selectProvisioningContext = JSON.parse(fs.readFileSync(p3, 'utf8'));

  return {
    contextForProvisioning: jest.fn().mockReturnValue(selectProvisioningContext),
  };
});

describe('Sync event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();
    ex = new FauxExpress();
  });

  it('All provisioned profile ids are returned', async () => {
    const req = {};
    client.query.mockReturnValueOnce({ rows: selectProfile });
    client.query.mockReturnValueOnce({ rows: selectProfileNamespaces });

    // @ts-ignore
    await getAllProvisionedProfileIds(req, ex.res);

    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch all provisioned profile ids should throw', async () => {
    const req = {
      params: {},
    };
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(getAllProvisionedProfileIds(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(ex.responseData).toBeUndefined();
  });

  it('Bot json object for a queried provisioned profile is returned', async () => {
    const req = {
      params: {
        profileId: 1,
      },
    };
    client.query.mockReturnValueOnce({ rows: selectProfile });
    client.query.mockReturnValueOnce({ rows: selectProfileNamespaces });

    // @ts-ignore
    await getProvisionedProfileBotJson(req, ex.res);

    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Bot json object for a queried provisioned profile should throw', async () => {
    const req = {
      params: { profileId: 1 },
    };
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(getProvisionedProfileBotJson(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(ex.responseData).toBeUndefined();
  });
});
