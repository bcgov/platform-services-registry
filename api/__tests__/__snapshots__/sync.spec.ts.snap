// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync event handlers All ids of profiles under pending edit / create are returned 1`] = `200`;

exports[`Sync event handlers All ids of profiles under pending edit / create are returned 2`] = `
Array [
  118,
]
`;

exports[`Sync event handlers All provisioned profile ids are returned 1`] = `200`;

exports[`Sync event handlers All provisioned profile ids are returned 2`] = `
Array [
  118,
]
`;

exports[`Sync event handlers Bot json object for a queried profile under pending edit / create is returned 1`] = `200`;

exports[`Sync event handlers Bot json object for a queried profile under pending edit / create is returned 2`] = `""`;

exports[`Sync event handlers Bot json object for a queried profile under pending edit / create should throw 1`] = `
"Unable to get profile (currently under pending edit / create) bot json
    for profile ID undefined"
`;

exports[`Sync event handlers Bot json object for a queried provisioned profile is returned 1`] = `200`;

exports[`Sync event handlers Bot json object for a queried provisioned profile is returned 2`] = `""`;

exports[`Sync event handlers Bot json object for a queried provisioned profile should throw 1`] = `"Unable get provisioned profile bot json for profile ID 1"`;

exports[`Sync event handlers Fetch all ids of profiles under pending edit / create should throw 1`] = `"Unable fetch all profile ids that are under pending edit"`;

exports[`Sync event handlers Fetch all provisioned profile ids should throw 1`] = `"Unable fetch all provisioned profile ids"`;
